Fix inbound traffic routing with Istio for VMI with explicitly specified ports

rhrazdil opened the pull request:
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>


**What this PR does / why we need it**:

### Add specific nat rules for inbound envoy routing
This commit adds nftable rules that correctly route traffic when Istio sidecar is injected to VMI.

Iptable rules are not being added intentionally for testability reasons - iptables are not tested in KubeVirt CI anymore.
`nftables` are required for this feature.

The Envoy proxy treats inbound traffic in the following manner:
- traffic targetting `port` declared in an associated k8s service: after proxying is sent from 127.0.0.1 to 127.0.0.1:`port`
- traffic targetting `port` NOT declared in an associated k8s service: after proxying is sent from 127.0.0.6 to `podIP:port` 

Changes:
- do not add dnat rule to KUBEVIRT_PREROUTING **if** Istio is used
- add rules to output and KUBEVIRT_POSTINBOUND chains that forward traffic that originates from 127.0.0.6


### e2e tests for VMI with Istio sidecar:

Added test scenarios that cover the followig cases for VMIs with
explicitly specified list of ports:

**1) Inbound traffic**
  - Test traffic is correctly routed to the VM
      - svc running on a port declared in k8s svc is reachable
      - svc running on a port NOT declared in k8s svc is reachable

  - With PeerAuthentication (Istio resource) set to only accept mTLS
    traffic (that is only from clients within the mesh).
    These tests verify that traffic is routed through the proxy
    and non-TLS encrypted traffic is dropped and doesn't reach VMI
      - svc running on a port declared in k8s svc is not reachable
      - svc running on a port NOT declared in k8s svc is not reachable


**2) Outbound traffic**
  - Test that VMI outside of service mesh is reachable
  - With OutboundTrafficPolicy set to REGISTRY_ONLY (that is, external
    services (anything outside mesh) must have explicit ServiceEntry
    allowing outbount traffic to them).
    This test verifies that outbound traffic is proxied and doesn't
    get through without
    - Test that VMI outside of service mesh is not reachable

Tests require to run with 1.20 and higher provider because the
option to deploy Istio was added in kubevirtci 1.20.

KUBEVIRT_DEPLOY_ISTIO env var must be to true to have Istio deployed.
Otherwise tests are skipped.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Design doc:
https://docs.google.com/document/d/1aBBCFBNxnOEQn3NH1ZiOcVnhHuiWmo3w451agmZxi_Q/edit

Important comments from the loong conversation
https://github.com/kubevirt/kubevirt/pull/5200#discussion_r611437879


**Release note**:

```release-note
Add support for Istio proxy traffic routing with masquerade interface. nftables is required for this feature.
```

rhrazdil commented:
nat rules with this patch for a VM with explicit ports 80, 8080 specified in VM interface:
```
table ip nat {
	chain PREROUTING {
		type nat hook prerouting priority dstnat; policy accept;
		iifname "k6t-eth0" counter packets 25 bytes 1814 jump ISTIO_REDIRECT
		iifname "k6t-eth0" counter packets 4 bytes 554 return
		meta l4proto tcp counter packets 178 bytes 10680 jump ISTIO_INBOUND
		iifname "eth0" counter packets 176 bytes 10560 jump KUBEVIRT_PREINBOUND
	}

	chain INPUT {
		type nat hook input priority 100; policy accept;
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		ip saddr 10.0.2.2 counter packets 2 bytes 142 masquerade
		oifname "k6t-eth0" counter packets 4 bytes 858 jump KUBEVIRT_POSTINBOUND
	}

	chain OUTPUT {
		type nat hook output priority -100; policy accept;
		meta l4proto tcp counter packets 30 bytes 1800 jump ISTIO_OUTPUT
		ip daddr 127.0.0.1 tcp dport 8080 counter packets 1 bytes 60 dnat to 10.0.2.2
		ip saddr 127.0.0.6 tcp dport 8080 counter packets 0 bytes 0 dnat to 10.0.2.2
		ip daddr 127.0.0.1 tcp dport 80 counter packets 0 bytes 0 dnat to 10.0.2.2
		ip saddr 127.0.0.6 tcp dport 80 counter packets 1 bytes 60 dnat to 10.0.2.2
	}

	chain ISTIO_INBOUND {
		meta l4proto tcp tcp dport 15008 counter packets 0 bytes 0 return
		meta l4proto tcp tcp dport 22 counter packets 0 bytes 0 return
		meta l4proto tcp tcp dport 15020 counter packets 0 bytes 0 return
		meta l4proto tcp tcp dport 15021 counter packets 176 bytes 10560 return
		meta l4proto tcp tcp dport 15090 counter packets 0 bytes 0 return
		meta l4proto tcp counter packets 2 bytes 120 jump ISTIO_IN_REDIRECT
	}

	chain ISTIO_REDIRECT {
		meta l4proto tcp counter packets 21 bytes 1260 redirect to :15001
	}

	chain ISTIO_IN_REDIRECT {
		meta l4proto tcp counter packets 3 bytes 180 redirect to :15006
	}

	chain ISTIO_OUTPUT {
		meta l4proto tcp tcp dport 15020 counter packets 2 bytes 120 return
		oifname "lo" ip saddr 127.0.0.6 counter packets 1 bytes 60 return
		oifname "lo" ip daddr != 127.0.0.1 skuid 1337 counter packets 1 bytes 60 jump ISTIO_IN_REDIRECT
		oifname "lo" skuid != 1337 counter packets 0 bytes 0 return
		skuid 1337 counter packets 26 bytes 1560 return
		oifname "lo" ip daddr != 127.0.0.1 skgid 1337 counter packets 0 bytes 0 jump ISTIO_IN_REDIRECT
		oifname "lo" skgid != 1337 counter packets 0 bytes 0 return
		skgid 1337 counter packets 0 bytes 0 return
		ip daddr 127.0.0.1 counter packets 0 bytes 0 return
		counter packets 0 bytes 0 jump ISTIO_REDIRECT
	}

	chain prerouting {
		type nat hook prerouting priority dstnat; policy accept;
	}

	chain input {
		type nat hook input priority 100; policy accept;
	}

	chain output {
		type nat hook output priority -100; policy accept;
	}

	chain postrouting {
		type nat hook postrouting priority srcnat; policy accept;
	}

	chain KUBEVIRT_PREINBOUND {
		tcp dport 8080 counter packets 0 bytes 0 dnat to 10.0.2.2
		tcp dport 80 counter packets 0 bytes 0 dnat to 10.0.2.2
	}

	chain KUBEVIRT_POSTINBOUND {
		tcp dport 8080 ip saddr 127.0.0.6 counter packets 0 bytes 0 snat to 10.0.2.1
		tcp dport 8080 ip saddr 127.0.0.1 counter packets 1 bytes 60 snat to 10.0.2.1
		tcp dport 80 ip saddr 127.0.0.6 counter packets 1 bytes 60 snat to 10.0.2.1
		tcp dport 80 ip saddr 127.0.0.1 counter packets 0 bytes 0 snat to 10.0.2.1
	}
}

```

rhrazdil commented:
/retest

rhrazdil commented:
/retest

alonSadan reviewed (approved):
Thank you Radim!

alonSadan commented:
I know I lgtmed already, but I thought of something. You hit this issue when there was no inbound traffic to the VM right? 
Is there a way to test this, or maybe there is already an existing test? To show that  it did solved the issue you faced.

rhrazdil commented:
> I know I lgtmed already, but I thought of something. You hit this issue when there was no inbound traffic to the VM right?
Is there a way to test this, or maybe there is already an existing test? To show that it did solved the issue you faced.

That's a good point, there is no test in place atm.
It can be tested by listening on port undeclared in a service, for example `echo -e "HTTP/1.1 200 OK\n\n Hello" | nc -l -p 8080`
and `curl <virt-launcher-pod-IP>:8080`.

AlonaKaplan commented:
Please add a release note.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1122,6 +1130,15 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 		if port.Protocol == "" {
 			port.Protocol = "tcp"
 		}
+		err = Handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_POSTINBOUND",
```
Shouldn't this rule be added outside the ports loop (only if no ports were declared on the vm)?
AFAIU the [scenario](https://github.com/istio/istio/issues/29603). If the port is declared on the pod container (in our case the virt-launcher compute container inherits ports declared on the vm/i) then 127.0.0.1 is used. Otherwise, 127.0.0.6 is used. In our case, we want to pass to the vm only ports that are declared on the vm or any port if nothing is declared. So from my understanding we should add the 127.0.0.6 rule only in case no ports are declared on the vm and it should forward any port.

rhrazdil replied:
> AFAIU the scenario. If the port is declared on the pod container (in our case the virt-launcher compute container inherits ports declared on the vm/i) then 127.0.0.1 is used. Otherwise, 127.0.0.6 is used. In our case, we want to pass to the vm only ports that are declared on the vm or any port if nothing is declared.

It dosn't depend on what ports are specified in our VMI/virt-launcher pod, it is the ports that are specified in an associated k8s service resource that makes the difference. If you look at this comment in that issue https://github.com/istio/istio/issues/29603#issuecomment-746878267
You'll see this comment describes the following scenario:

```
apiVersion: v1
kind: Service
metadata:
  name: vmi-cirros
spec:
  selector:
    app: vmi-cirros
  ports:
    - port: 8080
    
Let's assume VMI with the following interface:
      interfaces:
        - name: default
          masquerade: {}
          ports:
            - port: 8080
            - port: 8081


curl <virt-launcher-pod_IP>:8080 => Envoy proxies the packet from 127.0.0.1 to 127.0.0.1:8080
curl <virt-launcher-pod_IP>:8081 => Envoy proxies the packet from 127.0.0.6 to <virt-launcher-pod_IP>:8081
```

>  So from my understanding we should add the 127.0.0.6 rule only in case no ports are declared on the vm and it should forward any port.

We could add these two rules outside the ports loop as well, where we wouldn't match the destination port. It would fix this problem for a VM without explicit ports list too. I opted for adding rules that are as specific as possible.

AlonaKaplan replied:
Understood, thanks for the explanation.
Since this new rule and the next one are practically the same. I would add them in a for loop (iterating over the possible loopback addresses) instead of copy pasting.

rhrazdil replied:
> I would add them in a for loop

Done

AlonaKaplan commented:
I agree with Alon, an e2e test for istio is missing in general and therefore, for this specific use case as well.
Although the previous code was merged without tests, I believe we cannot continue the tradition.

rhrazdil commented:
@AlonaKaplan Added e2e tests for inbound traffic scenarios. I tested them locally, but since 1.20 and higher is required for them to run, these tests are currently not being executed in CI

rhrazdil commented:
rebased

AlonaKaplan commented on automation/test.sh:
```diff
@@ -50,6 +50,10 @@ elif [[ $TARGET =~ sig-network ]]; then
   export KUBEVIRT_PROVIDER=${TARGET/-sig-network/}
 elif [[ $TARGET =~ sig-storage ]]; then
   export KUBEVIRT_PROVIDER=${TARGET/-sig-storage/}
+  export KUBEVIRT_DEPLOY_ISTIO=true
```
Why did you place it in sig-storage?

rhrazdil replied:
ops, thanks!

rhrazdil replied:
Fixed

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
```
2021

rhrazdil replied:
Done

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+	}
+}
+
+func newRandomCirrosVMIWithIstioProxyAndEphemeralDiskAndUserDataAndAndServiceAccountDiskAndNetworkAndInterface(serviceAccount, appLabel, istioProxyAnnotationValue, userData string, network *v1.Network, iface *v1.Interface) *v1.VirtualMachineInstance {
```
Please use `libvmi.NewCirros`, to create the vm.

rhrazdil replied:
Done

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+	return vmi
+}
+
+var _ = FSIGDescribe("[Serial] Istio", func() {
```
Forgot to remove the F?

rhrazdil replied:
Indeed Thanks!

rhrazdil replied:
Done

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+			var srcVMI, dstVMI *v1.VirtualMachineInstance
+			switch destination {
+			case "vmiWithExplicitPorts":
+				srcVMI = vmiWithDefaultPorts
```
To make this test lighter, I would use a pod as the src instead of a vm. Is there a good reason to use a vm?

rhrazdil replied:
> Is there a good reason to use a vm?

My resoning is the following, I'm using both VMs in the 4 test cases, in 2 testcases, on acts as server and the other as a client, and in the other 2 test cases, it's the other way around. Since I need an associated service distinquished by a selector, I need 2 VMIs eventually.

I suppose I could serialize this so that for the first 2 cases I create 1 vmi  and for the other 2 cases I create another one. That would save 1 vmi worth of resources during the peak in exchange for time.

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+
+			By(fmt.Sprintf("checking that service running on port %d can be reached", port))
+			cmdCheck := fmt.Sprintf("curl %s:%d", dstVMIAddress, port)
+			err = console.SafeExpectBatch(srcVMI, []expect.Batcher{
```
`console.RunCommand` can be used here.

rhrazdil replied:
console.RunCommand is so much faster, thanks! Done

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+		},
+			table.Entry("declared port in VMI with explicit ports", "vmiWithExplicitPorts", svcDeclaredTestPort),
+			table.Entry("undeclared port in  VMI with explicit ports", "vmiWithExplicitPorts", svcUndeclaredTestPort),
+			table.Entry("declared port in VMI with default masquerade interface", "vmiWithDefaultPorts", svcUndeclaredTestPort),
```
`svcDeclaredTestPort` should be passed, isn't it?
And we know this scenario is not working AFAIU, so we need to add a skip and link to an issue.

rhrazdil replied:
Good catch! Copy paste errors are killing me here really.

The first commit in this PR fixes this case: `undeclared port in VMI with explicit ports`, without the iptable rule, the traffic isn't routed to the VM after it's proxied (because it's proxied from the 127.0.0.6). 

The last two cases do work in a way that the traffic does get into the VM, but it's not proxied. That is not being addressed in this PR. Ultimately though, we always want the traffic to reach the VM so IMO we should keep these cases.

rhrazdil replied:
Fixed

AlonaKaplan commented:
> Added e2e tests for inbound traffic scenarios. I tested them locally, but since 1.20 and higher is required for them to run, these tests are currently not being executed in CI


Thanks! Can you please explain in the commit message why 1.20 or higher k8s cluster is needed?

alonSadan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+			virtClient, err = kubecli.GetKubevirtClient()
+			tests.PanicOnError(err)
+
+			vmiWithExplicitPorts = newRandomCirrosVMIWithIstioProxyAndEphemeralDiskAndUserDataAndAndServiceAccountDiskAndNetworkAndInterface(
```
Maybe export  the vm creation to an helper? 
The two vm objects creations are almost a duplication.

rhrazdil replied:
I put some of the helper methods to utils, but not all of them as IMO some these are specific for istio scenario and it's unlikely that it would be used elsewhere.

alonSadan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+				_, err = virtClient.CoreV1().Services(svc.Namespace).Create(context.Background(), &svc, metav1.CreateOptions{})
+				Expect(err).ShouldNot(HaveOccurred())
+			}
+			for _, networkVMI := range []*v1.VirtualMachineInstance{vmiWithExplicitPorts, vmiWithDefaultPorts} {
```
Maybe merge this for-loop and the next one into on loop?

rhrazdil replied:
Done

alonSadan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+			}
+
+			// Wait for VMIs to become ready
+			vmiWithExplicitPorts = tests.WaitUntilVMIReady(vmiWithExplicitPorts, libnet.WithIPv6(console.LoginToCirros))
```
Can the waiting be also in a loop ?

rhrazdil replied:
Done

rhrazdil commented:
@AlonaKaplan  Address your comments in [force-push](https://github.com/kubevirt/kubevirt/compare/146aef8fc58c9c600253ef8edab3f48195c2bdec..ac90d163197fb9f80a720d663577167608b6f70b)

alonSadan reviewed (commented):
Thank you Radim!

rhrazdil commented:
@alonSadan Thanks for your comments, refactored the code in this [force-push](https://github.com/kubevirt/kubevirt/compare/321eb61ac4a1475dfab10d0ddf22e016ee4cfb41..f2407b4b6792c65eeccf0e2cca762903d24edd6c)

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+var _ = FSIGDescribe("[Serial] Istio", func() {
+	var err error
```
Thanks, done

rhrazdil commented:
/retest

rhrazdil commented:
/retest

alonSadan commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1136,24 +1136,17 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
-			"counter", "snat", "to", b.getGatewayByProtocol(proto))
-		if err != nil {
-			return err
+		for _, addr := range []string{getEnvoyInboundLoopbackAddress(proto), getLoopbackAdrress(proto)} {
```
I think this belongs to the previous commit.

rhrazdil replied:
Right! fixed

AlonaKaplan replied:
I don't like the coupling to envoy. Maybe we can match he entire loopback address block "127.0.0.0/8"?

alonSadan reviewed (commented):
I think It's great that  you exported `AddLabelToNamespace` to utils.
One  last note- I think one change belongs to the other commit.

rhrazdil commented:
/retest

AlonaKaplan commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1149,7 +1169,7 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 		}
 
 		err = Handler.NftablesAppendRule(proto, "nat", "output",
```
Why this new rule is needed?

AlonaKaplan replied:
when the source is 127.0.0.6 what is the dst address?

rhrazdil replied:
When src IP address is 127.0.0.6, then destination IP address is IP of the virt-launcher pod

rhrazdil replied:
Since I don't know what the virt-launcher pod IP is going to be (at least I think so) at this point in time, this rule is handling the `dnat` based on the source IP address.

rhrazdil replied:
Maybe we could consolidate the two rules to only one rule, which would not match the destination/source address, and only the destination port?

AlonaKaplan replied:
I don't like it since it becomes very specific to istio/envoy.
You can merge or keep it two rules but I would change it to check the port (as you mentioned) and the dst as well - we don't want to catch external destinations.
So I would change the original rule to daddr getLoopbackAdrress(proto) or podIp.

rhrazdil replied:
Updated accordingly to what we agreed on Thursday - keeping the istio annotation check.
> we don't want to catch external destinations.

I have changed the rules to avoid matching external destinations:

in output chain I match the following to `dnat`:
 - destinaton address is loopback
 - destination address is pod IP
 
and in postrouting chain I match only the VM address `10.0.2.2` to `snat` to `10.0.2.1`. I think this is a better solution, it safed us one rule in postrouting and also avoids matching external destinations. Also, port-forwarding should still work as port-forwarded traffic has src and dst IP loopback address

AlonaKaplan replied:
My comment was just about the output chain, so sounds great.

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,207 @@
+
+func newRandomCirrosVMIWithIstioSidecarAndDefaultServiceAccountAndNetworkAndInterface(appLabel string, network *v1.Network, ports []v1.Port) *v1.VirtualMachineInstance {
+	vmi := libvmi.NewCirros(
+		libvmi.WithNamespace(tests.NamespaceTestDefault),
```
You pass the namespace on the vmi creation, it is enough, there is no need to set in on the vmi as well.
@EdDev do you remember why it is even wrong to set the namespace on the vmi?

EdDev replied:
It is wrong because it may collide with the creation. When both are not identical, it will fail.
There is no need to specify the namespace in the spec before creation.

rhrazdil replied:
I see, removed this

AlonaKaplan commented on tests/libvmi/vmi.go:
```diff
@@ -71,6 +71,13 @@ func WithResourceMemory(value string) Option {
 }
 
+// WithNamespace specifies the vmi namespace.
+func WithNamespace(name string) Option {
```
Please refer to the comment regarding the namespace in `newRandomCirrosVMIWithIstioSidecarAndDefaultServiceAccountAndNetworkAndInterface`.

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,207 @@
+		withIstioAnnotation(),
+		withAppLabel(appLabel),
+	)
+	tests.AddServiceAccountDisk(vmi, "default")
```
Please add this (copied from your doc) or similar comment - " The serviceaccountdisk is required to have the service account token mounted, because it’s used by istio-proxy."

rhrazdil replied:
Added the comment

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,207 @@
+	}
+
+	Describe("Virtual machine in a service mesh using masquerade interface with explicit port list", func() {
+		tests.BeforeAll(func() {
```
We are trying to avoid the usage of `BeforeAll` in tests that are [not serial](https://github.com/kubevirt/kubevirt/pull/4051).
`BeforeAll` means that if each test will run on a different process the `BeforeAll` will be invoked as the number of the processes, which may be even less efficient than just having `BeforeEach` for every relevant test.

rhrazdil replied:
Refactored the test suite - I ditched BeforeAll in favor of BeforeEach and AfterEach.

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,207 @@
+			}
+			dstVMI, err := virtClient.VirtualMachineInstance(dstVMI.Namespace).Get(dstVMI.Name, &metav1.GetOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
+			dstVMIAddress := dstVMI.Status.Interfaces[0].IP
```
Missing ipv6 tests.
[Example](https://github.com/kubevirt/kubevirt/blob/876f1bec44659b8f3ee0784d6ac4a48c37f48a08/tests/vmi_networking_test.go#L668) of how to test ipv6.

rhrazdil replied:
IPv6 is not supported by Istio, so testing ipv6 is not relevant IMO

rhrazdil replied:
[done](https://github.com/kubevirt/kubevirt/compare/e47924b601c29ede8f8b175d6ede514fe9f480bb..2a20abaa3692d559d53ec3a034cd7465209461f0)

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,207 @@
+		},
+			table.Entry("declared port in VMI with explicit ports", vmiExplicitPortsAppCaseName, svcDeclaredTestPort),
+			table.Entry("undeclared port in  VMI with explicit ports", vmiExplicitPortsAppCaseName, svcUndeclaredTestPort),
+			table.Entry("declared port in VMI with default masquerade interface", vmiDefaultPortsAppCaseName, svcDeclaredTestPort),
```
I already commented about it but the comment is somewhere outdated.
I'm not sure how the last two scenarios are working. We don't have a code passing data from/to localhost to a vm with no specific ports (you call it - default masquerade interface), right? So how does this work?

rhrazdil replied:
As we discussed, added a [condition](https://github.com/kubevirt/kubevirt/compare/e47924b601c29ede8f8b175d6ede514fe9f480bb..2a20abaa3692d559d53ec3a034cd7465209461f0) to adding prerouting dnat rule only if istio annotation is not present.

e2e tests now exercise the following:
- inbound traffic can reach VM
- inbound traffic from outside the mesh cannot reach VM if Istio PeerAuthenticaton policy is set to strict, which verifies that inbound traffic is indeed being proxied
- outbound traffic can reach http://kubevirt.io
- outbound traffic can not reach http://kubevirt.io if Istio OutboundTrafficPolicy is set to REGISTRY_ONLY, which verifies that the outbound traffic is indeed being proxied

rhrazdil commented:
rebased

rhrazdil commented:
I found a bug in e2e test that tests outbound traffic, [fixed here](https://github.com/kubevirt/kubevirt/compare/d861a386d4e9c873f183dd665e00e6602964aa20..eb3c53ee59020c4c073130ef17e316fe79607c3f)

rhrazdil commented:
/retest

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+					&expect.BSnd{S: "\n"},
+					&expect.BExp{R: console.PromptExpression},
+					&expect.BSnd{S: "curl -sD - -o /dev/null http://kubevirt.io | head -n 1\n"},
+					&expect.BExp{R: fmt.Sprintf(".*%s.*\n", "HTTP/1.1 301 Moved Permanently")},
```
Are you sure this will always return this error code ? 

I just want to prevent us having to update the test if it starts returning 300 or 200 (or whatever other correct status code).

maiqueb replied:
You could extract it into a function `generatedExpectedHTTPStatusCodeString` or something alike.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+				err = console.SafeExpectBatch(vmi, []expect.Batcher{
+					&expect.BSnd{S: "\n"},
+					&expect.BExp{R: console.PromptExpression},
+					&expect.BSnd{S: "curl -sD - -o /dev/null http://kubevirt.io | head -n 1\n"},
```
This can be a constant. `curlKubevirtSite` or something alike.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+						return console.SafeExpectBatch(vmi, []expect.Batcher{
+							&expect.BSnd{S: "\n"},
+							&expect.BExp{R: console.PromptExpression},
+							&expect.BSnd{S: "curl -sD - -o /dev/null http://kubevirt.io | head -n 1\n"},
```
You can reuse the constant here.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+							&expect.BSnd{S: "\n"},
+							&expect.BExp{R: console.PromptExpression},
+							&expect.BSnd{S: "curl -sD - -o /dev/null http://kubevirt.io | head -n 1\n"},
+							&expect.BExp{R: fmt.Sprintf(".*%s.*\n", "HTTP/1.1 502 Bad Gateway")},
```
You could reuse the function mentioned above here.

rhrazdil replied:
Done

maiqueb commented on tests/utils.go:
```diff
@@ -3070,6 +3126,9 @@ func RenderPod(name string, cmd []string, args []string) *k8sv1.Pod {
 		ObjectMeta: metav1.ObjectMeta{
 			GenerateName: name,
 			Namespace:    NamespaceTestDefault,
+			Annotations: map[string]string{
+				IstioInjectSidecarAnnotation: "false",
+			},
```
I'm not sure how I feel about this. Why would all pods we use have to have this ?

rhrazdil replied:
I tried to address this in [this commit](https://github.com/kubevirt/kubevirt/pull/5200/commits/2667f5c2120ef2411079d0f3d239e2b0b2d418c1).
A simple alternative would be removing the label just before creating the Job, but that felt a little bit ackward.

maiqueb commented on tests/utils.go:
```diff
@@ -1809,6 +1813,58 @@ func createNamespaces() {
+	}
+	new.Labels[key] = value
+
+	newJson, err := json.Marshal(new)
+	Expect(err).ToNot(HaveOccurred())
+
+	patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, ns)
+	Expect(err).ToNot(HaveOccurred())
+
+	_, err = virtCli.CoreV1().Namespaces().Patch(context.Background(), ns.Name, types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func RemoveLabelFromNamespace(namespace, key string) {
+	virtCli, err := kubecli.GetKubevirtClient()
+	PanicOnError(err)
+
+	ns, err := virtCli.CoreV1().Namespaces().Get(context.Background(), namespace, metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	old, err := json.Marshal(ns)
+	Expect(err).ToNot(HaveOccurred())
+
+	new := ns.DeepCopy()
+	if new.Labels == nil {
+		// There are no labels, nothing to do
+		return
+	}
+	delete(new.Labels, key)
+
+	newJson, err := json.Marshal(new)
+	Expect(err).ToNot(HaveOccurred())
+
+	patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, ns)
+	Expect(err).ToNot(HaveOccurred())
+
+	_, err = virtCli.CoreV1().Namespaces().Patch(context.Background(), ns.Name, types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	Expect(err).ToNot(HaveOccurred())
+}
+
```
Is there an alternative where we can put this in the `test/network` package ? 

I'd rather not bloat the utils file with more stuff, especially if it can be avoided.

rhrazdil replied:
Before I go ahead and start being ceative, what do you think about `tests/network/utils/namespace.go`?

maiqueb replied:
Good enough for me. Let someone pitch in w/ the naming before moving it if you prefer.

maiqueb replied:
I want to correct myself; I actually meant something on [libnet](https://github.com/kubevirt/kubevirt/tree/1d7798a572eedcdb35f3292104006dd29db8dfe3/tests/libnet)

rhrazdil replied:
Moved to `tests/libnet/namespace.go`

maiqueb reviewed (changes requested):
Most of my comments are pretty much suggestions, are opinionated. As such, you can pretty much say you think I'm wrong because XYZ and I'm OK with it.

But I do feel the first commit should be better documented that it is in the commit msg. You have created a wonderful document that explains in detail what is happening in the pipeline. That information is paramount to understand what is happening. 

While I am not asking for you to translate the diagrams to ascii art ... It would be amazing :) 

Other than that, in the tests, there are plenty of checks that you can inline.

Very nice work here.

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+			)
+		}
+		BeforeEach(func() {
+			tests.BeforeTestCleanup()
```
This doesn't remove the namespaces, it only cleans bunch of resource types. So it should be pretty much no-op.
It shouldn't ever clean anything created by me, unless I'm missing something, so it's here only to clean potential leftovers.
I put it here because I saw this used in `vmi_networking.go` but as far as I'm concerned, this can be removed.

maiqueb replied:
It doesn't remove the namespaces, but it does delete pretty much everything left in them: https://github.com/kubevirt/kubevirt/blob/1ed6b509e32577c253c3cca2e83d04ff008bf965/tests/utils.go#L1650

This is nuking the playground after you're done with it. I'm just asking if this was what we've decided long ago. I think so, but I also remember an effort of cleaning up each entity - I don't remember the reasons why we've stopped doing that.

rhrazdil replied:
Removed the `tests.BeforeTestCleanup()` and added `AfterEach` for Inbound tests to delete the Job.

maiqueb replied:
I am unsure this is what we want to do... That was why I asked @AlonaKaplan what her view on this is, because I remember we've spoken about such things in the past, and have reached some sort of consensus. 

You might have to put this as it was, but for now, let's wait for her answer.

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1086,7 +1086,21 @@ func (b *MasqueradeBindMechanism) getVifIpByProtocol(proto iptables.Protocol) st
+	return false
+}
+
+func getIPForProtoVersion(proto iptables.Protocol, ipv4, ipv6 string) string {
```
I agree that the signature was odd, updated as you suggested

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1086,7 +1086,21 @@ func (b *MasqueradeBindMechanism) getVifIpByProtocol(proto iptables.Protocol) st
+	return ipv6
+}
+
+func getLoopbackAddress(proto iptables.Protocol) string {
```
Done

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+	return false
+}
+
+func withAppLabel(value string) libvmi.Option {
```
Done

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+
+func istioServiceMeshDeployed() bool {
+	value := os.Getenv(istioDeployedEnvVariable)
+	if value == "true" {
```
Actually this is a good point, not even that important here but in podinterface.go. Anyway, updated at both places

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+	}
+}
+
+func newServiceWithAppSelector(name, namespace, selector string) *corev1.Service {
```
Didn't find it anywhere, I can add it later when I deal with VM without explicit ports

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+		AfterEach(func() {
+			tests.RemoveLabelFromNamespace(tests.NamespaceTestDefault, istioInjectNamespaceLabel)
+			err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(vmi.Name, &metav1.DeleteOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
```
Done

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+			err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Delete(vmi.Name, &metav1.DeleteOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
+			err = virtClient.CoreV1().Services(tests.NamespaceTestDefault).Delete(context.TODO(), service.Name, metav1.DeleteOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
```
Done

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+		Describe("Inbound traffic", func() {
+			table.DescribeTable("request to VMI", func(port int) {
+				var job *batchv1.Job
+				job, err = verifyClientServerConnectivity(vmi, port)
```
As mentioned in this comment: https://github.com/kubevirt/kubevirt/pull/5200#discussion_r614976212
Added AfterEach to remove the Job with its children

AlonaKaplan replied:
Sorry I missed you tagged me on the comment. We decided to use the `BeforeTestCleanup`, since currently that's what most of out tests are using.
I don't remember the exact reason why removing just what you created in the after each caused us issued. @EdDev  do you remember?

EdDev replied:
It is unfortunate, but the current testing infrastructure we have reports the current state of the system (logs, entities states, etc) at a stage that occurs **after** the teardown is executed. As an outcome, on failure, cleanups that are done in the teardown will not be collected by the reporter and we will lack some information for debugging.

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,297 @@
+				By("Waiting for the job to succeed")
+				Expect(tests.WaitForJobToSucceed(job, 480*time.Second)).To(Succeed())
+			},
+				table.Entry(fmt.Sprintf("should reach VMI HTTP server on service declared port %d", svcDeclaredTestPort), svcDeclaredTestPort),
+				table.Entry(fmt.Sprintf("should reach VMI HTTP server on service undeclared port %d", svcUndeclaredTestPort), svcUndeclaredTestPort),
```
I guess you're right, I won't remember which port is which forever anyway, I'll remove it

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1033,7 +1033,7 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingIptables(protocol iptables.
 			strings.ToLower(port.Protocol),
```
Added the explanation to the commit message and elaborated on the inbound routing process

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			)
+		}
+		BeforeEach(func() {
+			libnet.AddLabelToNamespace(tests.NamespaceTestDefault, istioInjectNamespaceLabel, "enabled")
```
Please invoke `tests.BeforeTestCleanup()` to clean the lefovers of previous tests.

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			tests.WaitUntilVMIReady(vmi, console.LoginToCirros)
+		})
+		AfterEach(func() {
+			libnet.RemoveLabelFromNamespace(tests.NamespaceTestDefault, istioInjectNamespaceLabel)
```
`tests.BeforeTestCleanup()` should  remove vms and services.

rhrazdil replied:
Done

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			var job *batchv1.Job
+			AfterEach(func() {
+				var backgroundPropagation = metav1.DeletePropagationBackground
+				Expect(virtClient.BatchV1().Jobs(tests.NamespaceTestDefault).Delete(context.TODO(), job.Name, metav1.DeleteOptions{PropagationPolicy: &backgroundPropagation})).ShouldNot(HaveOccurred())
```
same. `tests.BeforeTestCleanup()` cleans the jobs.

rhrazdil replied:
Removed

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+				By("Waiting for the job to succeed")
+				Expect(tests.WaitForJobToSucceed(job, 480*time.Second)).To(Succeed())
+			},
+				table.Entry("should reach VMI HTTP server on service declared port", svcDeclaredTestPort),
```
How do we know for sure that envoy was involved? The fact that with peer authentication it doesn't work?

rhrazdil replied:
Yes

AlonaKaplan commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			})
+		})
+		Describe("Outbound traffic", func() {
+			var curlKubevirtioCommand = "curl -sD - -o /dev/null http://kubevirt.io | head -n 1\n"
```
I'm not sure we can assume we have external connectivity.
@EdDev I remember you wanted to send a PR adding a flag allowing testing external connectivity. Was there any progress with that?

@rhrazdil is there another way to test the outbound traffic without assuming the cluster has external connectivity.

EdDev replied:
External connectivity is supposed to be treated in this PR: https://github.com/kubevirt/kubevirt/pull/5169
It treats it by allowing the executor to mention if this is supported on a specific setup.

We probably need to do something beyond that even, by providing a ref point we setup ourself.

rhrazdil replied:
@AlonaKaplan Yes, I believe I can do that.
I didn't know that we should avoid external, note that kubevirt.io is used in vmi_network.go as well.

rhrazdil replied:
Done, I'm using another VMI that is not part of the mesh and acts as an external entity

AlonaKaplan commented on tests/job.go:
```diff
@@ -150,3 +158,15 @@ func NewHelloWorldJobHTTP(host string, port string, checkConnectivityCmdPrefixes
 	return NewJob("netcat", []string{"/bin/bash", "-c"}, []string{checkConnectivityCmd}, JobRetry, JobTTL, JobTimeout)
 }
+
+func NewHelloWorldJobHTTPWithDisabledIstioSidecar(host string, port string, checkConnectivityCmdPrefixes ...string) *batchv1.Job {
```
Maybe pass the annotations as vararg to the original methods instead of introducing new ones.

rhrazdil replied:
@AlonaKaplan Could you please elaborate?

AlonaKaplan replied:
https://github.com/kubevirt/kubevirt/pull/5529 is removing `checkConnectivityCmdPrefixes`, so you can add `annotations ...string`

rhrazdil replied:
Thank you. I ended up working around the issue, I realized that I only want to enable the sidecar in the upper-most BeforeEach, so I decided to label and de-label the namespace in this BeforeEach.

EdDev reviewed (commented):
The 2nd commit is vendoring istio client, but it is almost impossible to review. Could you please split it into the parts you directly changed and the ones that have been added as part of triggering the generated code and deps?

Also, please reason for adding this to the project. To me, it looks pretty problematic.
It is one thing that we support integration with istio, but a totally different thing that we need to import its codebase (which seems to be huge). Can we use a generic way here?

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1136,19 +1154,21 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 			"dport",
 			strconv.Itoa(int(port.Port)),
-			Handler.GetNFTIPString(proto), "saddr", getLoopbackAdrress(proto),
+			Handler.GetNFTIPString(proto), "daddr", b.getVifIpByProtocol(proto),
```
> Doesn't it mean that even if there is no istio at all, the source address of the packets arrive to the vm will be the ip of the bridge (10.0.2.1)?
I think it would be better to keep the original sending ip when possible.

Yes, you're right, the consequence of this is that the VM would see the origin as 10.0.2.1.

> BTW, how the opposite direction works? If the vm sends a reply who puts the original sender ip address on the packet? Envoy?

IIUC it's handled by the SNAT rule.
http://alumni.cse.ucsc.edu/~jed/articles/iptables/iptables-tutorial.html#RCTEST-IPTABLESTXT section 3.3 is the best explanation I could find, but let me try to elaborate.

So, the nat table is traversed only for the first packet of a new connection. As this table is traversed, an underlying nf_conntrack table is constructed. This table contains active connections (connection is there identified by: protocol, src/dst IP and src/dst port).
You can examine these active connections by looking at /proc/net/nf_conntrack
It contains something like this:
```
ipv4     2 tcp      6 119 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34374 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34374 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
ipv4     2 tcp      6 114 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34354 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34354 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
ipv4     2 tcp      6 99 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34252 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34252 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
```

These are just 3 curl request to a hello world nc server running in the VM issued from the node.
You can see that at the end of line, there are 2 uses of each connection - request + response.
The response is therefore not going through the nat table at all, it's source IP is determined by the SNAT table mapping from the request destination IP and port.

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1160,8 +1180,24 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
+			if err != nil {
+				return err
+			}
+			err = Handler.NftablesAppendRule(proto, "nat", "output",
```
> Why no just adding an or to the already existing output rule? Which, IIUC, exists only for istio so can be inside this if statement as well.

There is one catch, this current output rule is what makes port-forwarding work, if I put that under istio condition, I break port-forwarding. I don't think we want that as there is an effort to make port-forwarding work on VMs without explicit list of ports.

I could put both addresses under single rule, but then there would be both even without Istio, which I think is not desirable.

rhrazdil commented:
Rebased and addressed comments.
Important chenges:
- I have changed the rules added to output and KUBEVIRT_POSTINBOUND back to what I did before, because as @AlonaKaplan pointed out, matching the destination IP in KUBEVIRT_POSTINBOUND would cause the source address of the packets coming to the VM from pod network to be 10.0.2.1.

- Dropped afterEach calls and added cleanup of the new Istio resources to `BeforeTestCleanup`

- Dropped Istio client-go dependencies

- To test outbound traffic, I use another VMI that is not part of the mesh instead of using external kubevirt.io

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			)
+			service = newServiceWithAppSelector(
+				fmt.Sprintf("%s-service", vmi.Name),
+				vmi.Namespace,
```
Done

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+				{Port: svcUndeclaredTestPort},
+			}
+			vmi = newRandomCirrosVMIWithIstioSidecarAndDefaultServiceAccountAndNetworkAndInterface(
+				vmiAppSelector,
```
Done - I only pass port list, which is used

rhrazdil replied:
Ah sorry, now I see it's not used in this PR. However, I'll need the ports to be parametrized when working with VM without explicit list of ports.

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,310 @@
+			service = newServiceWithAppSelector(
+				fmt.Sprintf("%s-service", vmi.Name),
+				vmi.Namespace,
+				vmiAppSelector,
```
Done, not passing any parameters

maiqueb commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,99 @@
+	"kubevirt.io/client-go/kubecli"
+)
+
+func AddLabelToNamespace(namespace, key, value string) {
```
First of, thanks for exporting these.

Despite that, there's one thing I don't like - the expects inside this helper. I think you should return the error, and the caller is the one that should expect the call to succeed.

rhrazdil replied:
Thanks, I agree, I forgot to update these appropriately. Done

maiqueb commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,99 @@
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func RemoveLabelFromNamespace(namespace, key string) {
```
Same for this function.

rhrazdil replied:
Done

maiqueb commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,99 @@
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func RemoveAllLabelsFromNamespace(namespace string) {
```
ditto.

rhrazdil replied:
Done

maiqueb commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,99 @@
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func PanicOnError(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
```
I don't think you need to copy this function here; just receive the kubevirtClient as a parameter to all the functions that need it.

It might be a bit cumbersome, but, it is what we're doing everywhere else (unless I'm mistaken).

rhrazdil replied:
Updated as you suggested

maiqueb commented on tests/utils.go:
```diff
@@ -1757,6 +1764,26 @@ func cleanNamespaces() {
 			PanicOnError(virtCli.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(namespace).Delete(context.Background(), netDef.GetName(), metav1.DeleteOptions{}))
 		}
+
+		// Remove Istio Sidecars
+		sidecarResource := schema.GroupVersionResource{Group: "security.istio.io", Version: "v1beta1", Resource: "peerauthentications"}
+		sidecars, err := virtCli.DynamicClient().Resource(sidecarResource).Namespace(namespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			PanicOnError(err)
+		}
+		for _, sidecar := range sidecars.Items {
+			PanicOnError(virtCli.DynamicClient().Resource(sidecarResource).Namespace(namespace).Delete(context.Background(), sidecar.GetName(), metav1.DeleteOptions{}))
+		}
```
You could wrap this in an helper function and simply execute it here.

rhrazdil replied:
Done

maiqueb commented on tests/utils.go:
```diff
@@ -1757,6 +1764,26 @@ func cleanNamespaces() {
+			PanicOnError(virtCli.DynamicClient().Resource(sidecarResource).Namespace(namespace).Delete(context.Background(), sidecar.GetName(), metav1.DeleteOptions{}))
+		}
+
+		// Remove Istio PeerAuthentications
+		peerAuthenticationResource := schema.GroupVersionResource{Group: "security.istio.io", Version: "v1beta1", Resource: "peerauthentications"}
+		peerAuthentications, err := virtCli.DynamicClient().Resource(peerAuthenticationResource).Namespace(namespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			PanicOnError(err)
+		}
+		for _, pa := range peerAuthentications.Items {
+			PanicOnError(virtCli.DynamicClient().Resource(peerAuthenticationResource).Namespace(namespace).Delete(context.Background(), pa.GetName(), metav1.DeleteOptions{}))
+		}
```
ditto.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+	svcUndeclaredTestPort        = 1501
+)
+
+var (
+	vmi     *v1.VirtualMachineInstance
+	service *corev1.Service
+	nad     *k8snetworkplumbingwgv1.NetworkAttachmentDefinition
+)
```
I don't know how I didn't notice this earlier; why do you have these test globally reachable variables ? Can't we scope them properly, and use them as parameters to functions when required ?

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+				Spec: k8snetworkplumbingwgv1.NetworkAttachmentDefinitionSpec{},
+			}
+			nad, err = virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(tests.NamespaceTestDefault).Create(context.TODO(), nad, metav1.CreateOptions{})
+			ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
Do we need the offset ? afaiu, this would be useful if it were defined within an helper - something I try to avoid (having expectations inside helpers).

rhrazdil replied:
I used it as per this comment: https://github.com/kubevirt/kubevirt/pull/5200#discussion_r616985422
I'm changing it back to regular Expect as I don't see the point of the offset here either. 
@AlonaKaplan Could you please explain why should we use ExpectWithOffset here?

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+
+			By("Creating VMI and associated Service")
+			_, err = virtClient.CoreV1().Services(tests.NamespaceTestDefault).Create(context.Background(), service, metav1.CreateOptions{})
+			ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
ditto.

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+			_, err = virtClient.CoreV1().Services(tests.NamespaceTestDefault).Create(context.Background(), service, metav1.CreateOptions{})
+			ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
+			vmi, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
+			ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
ditto.

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+				peerAuthenticationRes := schema.GroupVersionResource{Group: "security.istio.io", Version: "v1beta1", Resource: "peerauthentications"}
+				var peerAuthentication *unstructured.Unstructured
+				BeforeEach(func() {
+					peerAuthentication = &unstructured.Unstructured{
+						Object: map[string]interface{}{
+							"apiVersion": "security.istio.io/v1beta1",
+							"kind":       "PeerAuthentication",
+							"metadata": map[string]interface{}{
+								"name": "strict-pa",
+							},
+							"spec": map[string]interface{}{
+								"mtls": map[string]interface{}{
+									"mode": "STRICT",
+								},
+							},
+						},
+					}
```
It's a bit distracting to see this here. Could you consider having it in an helper ? It could even be a lambda in this context - if this is not reused in other scopes. `generatePeerAuthRequest` or something alike ?

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+						},
+					}
+					_, err = virtClient.DynamicClient().Resource(peerAuthenticationRes).Namespace(tests.NamespaceTestDefault).Create(context.Background(), peerAuthentication, metav1.CreateOptions{})
+					ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
same comment as above w/ the offset.

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+					By("Waiting for the job to fail")
+					Expect(tests.WaitForJobToSucceed(job, 480*time.Second)).NotTo(Succeed())
+
+					err = virtClient.DynamicClient().Resource(peerAuthenticationRes).Namespace(tests.NamespaceTestDefault).Delete(context.Background(), peerAuthentication.GetName(), metav1.DeleteOptions{})
+					ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
I think you should move this teardown to an `AfterEach` block.

rhrazdil replied:
I removed the teardown from here completely and instead I clean the istio resources in BeforeTestCleanup.
As per https://github.com/kubevirt/kubevirt/pull/5200#discussion_r617235464
we should avoid using AfterEach because if error occurres in AfterEach, it doesn't get collected

maiqueb replied:
OK, I had missed the fact that you're removing it in https://github.com/kubevirt/kubevirt/pull/5200/files?file-filters%5B%5D=.go#diff-847cd06c1bf6ef597ea96894f8b32ecb4d074ad1bc785b39871e11fb6ab71b3dR1774. 

But anyway, I don't think this argument is that strong, since you're not collecting the logs of this type of entity (you need to update the k8s reporter for that to happen: https://github.com/kubevirt/kubevirt/blob/master/tests/reporter/kubernetes.go#L91).

Anyway, not trying to change your mind, it's good as it is (probably better :) ); just saying that since you're not dumping this info, you **could** have used an `AfterEach` .

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+			tests.WaitUntilVMIReady(vmi, console.LoginToCirros)
+		})
+		Describe("Inbound traffic", func() {
+			var job *batchv1.Job
```
do you really need the `job` as a variable defined in this scope ? 

afaict, it could be locally scoped anytime you create it (since you do not have to clean it up manually in the test tear down.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+		}
+	})
+	var (
+		err        error
```
is this really needed ? can't you just assign the error to a new variable whenever needed ? Anyway, not a must - I just find this off-putting.

rhrazdil replied:
I somehow prefer this instead of keeping track of where was the first err with `:=` and then making sure not to overshadow.

maiqueb replied:
I see it the other way around; errors are totally disposable, meant to be returned / printed once. If possible, they should be neatly in their scope, but if needed, well, nothing bad with shadowing them. 

Have it your way, this is a personal preference I can't impose on you.

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+				} else {
+					codes = append(codes, "5..")
+				}
+				return fmt.Sprintf("HTTP\\/[123456789\\.]{1,3}\\s(%s)", strings.Join(codes, "|")), nil
```
oh boy 0_o 

I start to regret having suggested this function . 

Can't say this is easy to read.

rhrazdil replied:
Hopefully, it's better now

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+
+				By("Getting back the Server VMI IP")
+				serverVMI, err = virtClient.VirtualMachineInstance(serverVMI.Namespace).Get(serverVMI.Name, &metav1.GetOptions{})
+				ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
the offset is not needed.

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+				return fmt.Sprintf("curl -sD - -o /dev/null http://%s:%d | head -n 1\n", serverIP, port)
+			}
+
+			generateExpectedHTTPReturnCodesRegex := func(shouldBeReachable bool) (string, error) {
```
Having a boolean argument is not a good thing; it's a clear indication your function does 2 things.

Wouldn't it simplify your code if you had 2 regexes ?

rhrazdil replied:
Yeah this wasn't good. I went ahead to make 2 regexes, but then I didn't like that it's only the code that's different.
So I've parametrized it by the return code instead. Should make more sense  this way

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+			Context("With Sidecar allowing only registered external services", func() {
+				var registryOnlySidecar *unstructured.Unstructured
+				BeforeEach(func() {
+					registryOnlySidecar = &unstructured.Unstructured{
+						Object: map[string]interface{}{
+							"apiVersion": "networking.istio.io/v1beta1",
+							"kind":       "Sidecar",
+							"metadata": map[string]interface{}{
+								"name": "registry-only-sidecar",
+							},
+							"spec": map[string]interface{}{
+								"outboundTrafficPolicy": map[string]interface{}{
+									"mode": "REGISTRY_ONLY",
+								},
+							},
+						},
+					}
```
consider also a generator function to return this `&unstructured.Unstructured` struct.

rhrazdil replied:
Done

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,313 @@
+						},
+					}
+					_, err = virtClient.DynamicClient().Resource(sidecarRes).Namespace(tests.NamespaceTestDefault).Create(context.TODO(), registryOnlySidecar, metav1.CreateOptions{})
+					ExpectWithOffset(1, err).ShouldNot(HaveOccurred())
```
with offset is not needed.

rhrazdil replied:
https://github.com/kubevirt/kubevirt/pull/5200#discussion_r621903611

maiqueb reviewed (changes requested):
While it might seem I got a ton of comments, I'm happy with how the tests look like.

Pretty much all of them are minor.

maiqueb commented on tests/utils.go:
```diff
@@ -1757,6 +1765,14 @@ func cleanNamespaces() {
 		}
+
+		// Remove all Istio Sidecars
+		err = removeAllGroupVersionResourceFromNamespace(schema.GroupVersionResource{Group: "networking.istio.io", Version: "v1beta1", Resource: "sidecars"}, namespace)
+		PanicOnError(err)
```
You can inline this; the err is not being used anywhere.

rhrazdil replied:
[Done](https://github.com/kubevirt/kubevirt/compare/ef7c453daccb75fb733dd6ef2c13636f0d5a51f7..5e074b4b40d1696b58a3c12d0802761156fcabe1)

maiqueb commented on tests/utils.go:
```diff
@@ -1757,6 +1765,14 @@ func cleanNamespaces() {
+		PanicOnError(err)
+
+		// Remove all Istio PeerAuthentications
+		err = removeAllGroupVersionResourceFromNamespace(schema.GroupVersionResource{Group: "security.istio.io", Version: "v1beta1", Resource: "peerauthentications"}, namespace)
+		PanicOnError(err)
```
ditto.

rhrazdil replied:
[Done](https://github.com/kubevirt/kubevirt/compare/ef7c453daccb75fb733dd6ef2c13636f0d5a51f7..5e074b4b40d1696b58a3c12d0802761156fcabe1)

maiqueb commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,304 @@
+			// Istio envoy may intercept the request, we need to check the header response, not just the
+			// return code of curl, because if envoy proxy forbids a request, it responds with 502 Bad Gateway code
+			It("Should be able to reach kubevirt.io", func() {
+				err = console.SafeExpectBatch(vmi, []expect.Batcher{
```
This is what I meant with the error comment; this `err` could be locally scoped. 

You don't need to do it, I'm just pointing it out.

maiqueb reviewed (approved):
Thank you.

All the comments I left are really just minor opinionated things that you can just pretend you didn't read.

I think the new tests look amazing, good work.

rhrazdil commented:
Most of the failed lanes don't seem to be related to changes in this PR, moreover, some tests results are not accessible as prow never loads up the output.
/retest

rhrazdil commented:
/test pull-kubevirt-e2e-k8s-1.19-operator

maiqueb commented on tests/utils.go:
```diff
@@ -1785,6 +1799,30 @@ func removeNamespaces() {
+
+	gvr, err := virtCli.DynamicClient().Resource(groupVersionResource).Namespace(namespace).List(context.Background(), metav1.ListOptions{})
+	if errors.IsNotFound(err) {
+		// gvr is not found, no-op
```
I think there's hardly any motivation for this comment; it's pretty obvious what you're doing.

No need to update the PR if it's just this; only consider it when another reviewer complains about something.

rhrazdil commented:
Rebased

AlonaKaplan commented:
> The annotation added to RenderPod in 2a20aba#diff-847cd06c1bf6ef597ea96894f8b32ecb4d074ad1bc785b39871e11fb6ab71b3dR3127
> is added to disable injecting Istio sidecar to these pods, because I noticed that using sidecar with these pods
> never works, presumably because the pod issues the request at same the moment Istio is manipulating iptables.

Couldn't find this in the current version of the code.

AlonaKaplan commented:
The changes are done only on nftables. Does the PR skip iptables intentionally? If so, please mention it in the PR message.

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1206,13 +1217,15 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
+			}
 		}
 
 		err = b.handler.NftablesAppendRule(proto, "nat", "output",
```
> I believe this rule was originally added for istio

I think you're correct. But, there is currently an [effort](https://github.com/kubevirt/kubevirt/pull/5431) to make kubectl port-forward work for VMIs.
While that effort is unrelated to this, coincidentally, this rule is needed for both port-forward and istio proxy to work.
(Traffic that is port-forwarded is pushed from loopback 127.0.0.1 destined to 127.0.0.1.)

For that reason, I think we want to keep these present

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1224,8 +1237,28 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 
+		if hasIstioSidecarInjectionEnabled(b.vmi) && proto == iptables.ProtocolIPv4 {
+			err = b.handler.NftablesAppendRule(proto, "nat", "output",
+				b.handler.GetNFTIPString(proto), "saddr", getEnvoyLoopbackAddress(),
```
I'm not sure I get the suggested rule, didn't you mean this?
```
err = b.handler.NftablesAppendRule(proto, "nat", "output",
b.handler.GetNFTIPString(proto), "saddr", "{" + getLoopbackAdrress(proto) + "," + getEnvoyLoopbackAddress() + "}",
			strings.ToLower(port.Protocol),
			"dport",
			strconv.Itoa(int(port.Port)),
			"counter", "dnat", "to",  b.getVifIpByProtocol(proto))
```
If so, than the reason is the same as in the above comment.
But what I could do here is to wrap `"{" + getLoopbackAdrress(proto) + "," + getEnvoyLoopbackAddress() + "}"` in a function and return what's appropriate:
{ getLoopbackAdrress } if Istio is not used
{ getLoopbackAdrress(), getEnvoyLoopbackAddress() } if Istio is used

What do you think?

AlonaKaplan replied:
> What do you think?

Yes, IMO it will be more readable.

rhrazdil replied:
[Done](https://github.com/kubevirt/kubevirt/pull/5200/commits/f3b5ebdbfa8098c948342a51d05a83d9db09ee1b)

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1206,13 +1217,15 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 			return err
```
See https://github.com/kubevirt/kubevirt/pull/5200/files/5cecb6dd0ba5cff680352e9d28387f55842eb26e#r626481338

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1224,8 +1237,28 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
+			if err != nil {
+				return err
+			}
+			err = b.handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_POSTINBOUND",
```
Similarly, please see https://github.com/kubevirt/kubevirt/pull/5200/files/5cecb6dd0ba5cff680352e9d28387f55842eb26e#r626496232

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,304 @@
+
+			// Enable sidecar injection by setting the namespace label
+			Expect(libnet.AddLabelToNamespace(virtClient, tests.NamespaceTestDefault, tests.IstioInjectNamespaceLabel, "enabled")).ShouldNot(HaveOccurred())
+			defer func() {
```
> EDIT: oh, you create the vm in the BeforeEach so I guess the label is not needed anymore after the vm is created. Anyway, I find it confusing.

Yes, I only add the label when creating the VMI that's to be tested. 

 > EDIT2: You've added the label removal to the general cleanup method. Why this specific cleanup is needed as well?
 
It shouldn't be needed. I added the removal to be extra sure that the label doesn't leak to other tests somehow.
But it shouldn't since I'm deferring the removal

rhrazdil commented on tests/network/vmi_istio.go:
```diff
@@ -0,0 +1,304 @@
+				// Envoy responds with 502 Bad Gateway return code.
+				// After Sidecar with OutboundTrafficPolicy is created, it may take a while for the Envoy proxy
+				// to sync with the change, first request may still get through, hence the Eventually.
+				It("Should not be able to reach kubevirt.io", func() {
```
Fixed

rhrazdil commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,114 @@
+	return nil
+}
+
+func RemoveAllLabelsFromNamespace(client kubecli.KubevirtClient, namespace string) error {
```
[Done](https://github.com/kubevirt/kubevirt/pull/5200/commits/a52e24d9494c559117587c749af81993bf3efa39)

rhrazdil commented:
> The changes are done only on nftables. Does the PR skip iptables intentionally? If so, please mention it in the PR message.

Updated PR description

rhrazdil commented:
/retest

AlonaKaplan commented:
> The changes are done only on nftables. Does the PR skip iptables intentionally? If so, please mention it in the PR message.

I still don't see in the PR description explanation why iptables weren't updated as well.

AlonaKaplan commented:
> > The changes are done only on nftables. Does the PR skip iptables intentionally? If so, please mention it in the PR message.
> 
> I still don't see in the PR description explanation why iptables weren't updated as well.

Basically approved. Just waiting to understand why iptables were skipped.

rhrazdil commented:
@AlonaKaplan Updated PR description and release note

maiqueb commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1113,6 +1113,18 @@ func (b *MasqueradeBindMechanism) getVifIpByProtocol(proto iptables.Protocol) st
 	}
 }
 
+func (b *MasqueradeBindMechanism) getDstAddressesToDnat(proto iptables.Protocol) (string, error) {
```
Could you define this function after the caller ?

rhrazdil replied:
Moved both of these, Thanks

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1246,20 +1246,19 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 	return nil
 }
 
+func getSrcAddressesToSnat(vmi *v1.VirtualMachineInstance, proto iptables.Protocol) string {
```
No reason in particular, I didn't add the masqueradeBind receiver because it's not used.
Similarly, `getLoopbackAddress` doesn't have it, although I see that this method is more specific for MasqueradeBind.
Do you think it should be added to express the relation?

maiqueb replied:
Well, you don't use the receiver, but you could - instead of using the vmi input parameter, you could use the member of the receiver struct.

Anyway. These are peanuts.

rhrazdil replied:
That's a good point actually, I missed that. I like the idea of these method having the same parameters

rhrazdil replied:
[done](https://github.com/kubevirt/kubevirt/compare/af0422b4fea220c2fe7c0a639d4b60844018199d..beac357b4698e4f150480521e6303d22bd2e30ae)

maiqueb commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1187,13 +1187,15 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
-		if err != nil {
-			return err
+		if !hasIstioSidecarInjectionEnabled(b.vmi) {
+			err = b.handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_PREINBOUND",
+				strings.ToLower(port.Protocol),
+				"dport",
+				strconv.Itoa(int(port.Port)),
+				"counter", "dnat", "to", b.getVifIpByProtocol(proto))
+			if err != nil {
+				return err
+			}
```
Would you export this to a function ? `performDNATOnPreInbound` or something like that ?

maiqueb commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1205,8 +1207,28 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 
+		if hasIstioSidecarInjectionEnabled(b.vmi) && proto == iptables.ProtocolIPv4 {
+			err = b.handler.NftablesAppendRule(proto, "nat", "output",
+				b.handler.GetNFTIPString(proto), "saddr", getEnvoyLoopbackAddress(),
```
Same here; would you export this to a function ? `performDNATForEnvoyTrafficOnOutputChain` or something similar.

Everytime you rebase, I have to read this again, and re-learn what this does. I think encapsulating this in a function with a meaningful name would be helpful.

Also for the SNAT performed below.

maiqueb commented on tests/libnet/namespace.go:
```diff
@@ -0,0 +1,114 @@
+	"encoding/json"
+
+	. "github.com/onsi/gomega"
+
```
This import is fishy; helpers shouldn't use gomega / ginkgo.

Why would you need matchers on this file ? I see that you remove the import in the next commit; please squash that 'removal' in this one.

rhrazdil replied:
There were forgotten `Expect` calls by mistake which is why the import remained.
I have [squashed](https://github.com/kubevirt/kubevirt/compare/2f31699ae8a76b1606c62c82c3f7fa3379eb52ac..622b1e4048c0f95f1c28026c45e26480e9f7ff97) the e2e test commit with the commit that refactored the namespace helpers - there's no need to have that commit separated

maiqueb reviewed (changes requested):
OK; I now see that my comments about exporting the SNAT/DNAT operations into functions no longer apply, since you know handle them differently. 

You can ignore those comments, and focus on the matchers (gomega) one instead.

maiqueb reviewed (approved):
Good to go !
